### PR TITLE
don't close server socket if fcntl(O_NONBLOCK) throws EINVAL

### LINKED ISSUE #1030: ServerSocketChannel behaves non-sensical if fcntl on accepted socket returns EINVAL
As of the docs, `fcntl` on a newly accepted socket shouldn't return `EINVAL` but it seems to sporadically do that on macOS. We need to analyse this. Still, it's probably not a sane behaviour that if that happens, we just close the server socket...

### Expected behavior

don't actually know but just silently closing the server socket isn't good

### Actual behavior

debug mode: crash
release mode: silently close server socket

### Steps to reproduce

dunno

### If possible, minimal yet complete reproducer code (or URL to code)

n/a

### SwiftNIO version/commit hash

all (probably)

### Swift & OS version (output of `swift --version && uname -a`)

macOS

### PR DESCRIPTION
**Pull request by weissi**
Motivation:

Darwin does sometimes hand us back an (as of the docs) illegal error
code of EINVAL. So far, NIO has done the non-sensical thing of closing
the server socket as a result.

Modifications:

Just close the accepted socket, leave the server socket open, and fire
the error through the pipeline.

Result:

- fixes #1030
- better behaviour

### CODE REVIEW

File: Sources/NIO/SocketChannel.swift [hunk 1]
```diff
@@ -436,6 +438,16 @@ final class ServerSocketChannel: BaseSocketChannel<ServerSocket> {
             // These are errors we may be able to recover from. The user may just want to stop accepting connections for example
             // or provide some other means of back-pressure. This could be achieved by a custom ChannelDuplexHandler.
             return false
+        case EINVAL:
+            if case .function(let function) = err.reason {
+                // see https://github.com/apple/swift-nio/issues/1030
+                // on Darwin, fcntl(fd, F_SETFL, O_NONBLOCK) sometimes returns EINVAL...
+                //
+                // the comparison is so weird because StaticString isn't Equatable and doesn't have an == method...
+                return String(describing: function) != "fcntl(descriptor:command:value:)"
```
Comment (Lukasa) [inline Sources/NIO/SocketChannel.swift]: Yuck, this seems like a really easy way to break the code later on. Can we push this error handling deeper into the code in any way? Potentially by catching the fcntl EINVAL and either rethrowing or doing something else? I kinda hate this string-based coupling.

Comment (Lukasa) [reply Sources/NIO/SocketChannel.swift]: In fact, maybe `accept` should swallow this.

Comment (weissi) [reply Sources/NIO/SocketChannel.swift]: @Lukasa I think we should deliver it

Comment (Lukasa) [reply Sources/NIO/SocketChannel.swift]: In that case, `accept` should wrap this error into something clearer.

Comment (weissi) [reply Sources/NIO/SocketChannel.swift]: @Lukasa agreed, implemented that yesterday but github still shows the 'outdated' change :|

File: Sources/NIO/Channel.swift [hunk 2]
```diff
@@ -349,6 +349,8 @@ public enum ChannelError: Error {
 
 extension ChannelError: Equatable { }
 
+public struct NIOFailedToSetSocketNonBlockingError: Error {}
```
Comment (Lukasa) [inline Sources/NIO/Channel.swift]: I think this needs a doc comment.

**Pull request merged**